test/hasTabs

```diff
@@ -56,9 +56,11 @@ then
   echo "Run the following command(s) on the above files to replace your TABs"
   echo "with four white spaces:"
   echo ""
+  echo "GNU_SED=\$(sed --help >/dev/null 2>&1 && { echo 1; } || { echo 0; })"
+  echo "[[ \${GNU_SED} -eq 1 ]] && REPLACE=\"sed -i 's/\t/\ \ \ \ /g'\" || REPLACE=\"sed -i '' -E 's/\$(printf '\t')/\ \ \ \ /g'\""
```

Comment:
This evaluetes to:
```
sed -i '' -E 's/ /\ \ \ \ /g'
```
on MacOS and thus does not replace tabs.

Reply:
```suggestion
  echo "[[ \${GNU_SED} -eq 1 ]] && REPLACE=\"sed -i 's/\t/\ \ \ \ /g'\" || REPLACE=\"sed -i '' -E \\\"s/\\\$(printf '\t')/\ \ \ \ /g\\\"\""
```

Reply:
With these changes, it works on my Mac.

Reply:
Wait, how did you check this? Please execute  the script and copy past what it prints on the terminal, as this is run in travis-ci.

We tested this in our repo and the suggested changes worked, as far as I remember. Just double checking how you executed this to be sure.

Reply:
That is exactly what I tested - I will do it again and print the output below:

Reply:
original text
```
this is a test text ending in many white spaces      
This is a text with a	tab.
```
in detail `od -t a tmp` leads to
```
0000000    t   h   i   s  sp   i   s  sp   a  sp   t   e   s   t  sp   t
0000020    e   x   t  sp   e   n   d   i   n   g  sp   i   n  sp   m   a
0000040    n   y  sp   w   h   i   t   e  sp   s   p   a   c   e   s  sp
0000060   sp  sp  sp  sp  sp  nl   T   h   i   s  sp   i   s  sp   a  sp
0000100    t   e   x   t  sp   w   i   t   h  sp   a  ht   t   a   b   .
0000120   nl                                                            
0000121
```

Execute: `GNU_SED=$(sed --help >/dev/null 2>&1 && { echo 1; } || { echo 0; })`
(result of `$GNU_SED` is `0`)

Execute
```
[[ ${GNU_SED} -eq 1 ]] && REPLACE="sed -i 's/\t/\ \ \ \ /g'" || REPLACE="sed -i '' -E 's/$(printf '\t')/\ \ \ \ /g'"
```
as copied from above.
`echo $REPLACE` returns `sed -i '' -E 's/ /\ \ \ \ /g'` - which is not the intended output for OS X because the `printf` is evaluated before.

Running `eval "${REPLACE}"  tmp` (`tmp` is my file name) leads to the following text:
```
this is a test text ending in many white spaces      
This is a text with a    tab.
```
and `od -t a` reveals:
```
0000000    t   h   i   s  sp   i   s  sp   a  sp   t   e   s   t  sp   t
0000020    e   x   t  sp   e   n   d   i   n   g  sp   i   n  sp   m   a
0000040    n   y  sp   w   h   i   t   e  sp   s   p   a   c   e   s  sp
0000060   sp  sp  sp  sp  sp  nl   T   h   i   s  sp   i   s  sp   a  sp
0000100    t   e   x   t  sp   w   i   t   h  sp   a  sp  sp  sp  sp   t
0000120    a   b   .  nl                                                
0000124
```
Thus the tab has been changed.

Reply:
However, the issue is caused by the quatation marks: in your post above you added them:
```diff
+eval "${REPLACE}"  <file1>
-eval ${REPLACE}  <file1>
```
but when I execute `echo "eval \${REPLACE} $i"` from [here(https://github.com/ComputationalRadiationPhysics/picongpu/blob/3b10f00ee45fa9e9203214867aafa3ab3f19f54c/test/hasEOLwhiteSpace#L65)] on OS X or any other system, echo removed the quotation marks `"`. 

If I then evaluate only that what is actually printed: 
```
eval ${REPLACE}  <file1>
```
all apaces are replaced by for spaces leading to the following text (again `od -t a` output after the text):
```
this    is    a    test    text    ending    in    many    white    spaces                        
This    is    a    text    with    a	tab.
```
```
0000000    t   h   i   s  sp  sp  sp  sp   i   s  sp  sp  sp  sp   a  sp
0000020   sp  sp  sp   t   e   s   t  sp  sp  sp  sp   t   e   x   t  sp
0000040   sp  sp  sp   e   n   d   i   n   g  sp  sp  sp  sp   i   n  sp
0000060   sp  sp  sp   m   a   n   y  sp  sp  sp  sp   w   h   i   t   e
0000100   sp  sp  sp  sp   s   p   a   c   e   s  sp  sp  sp  sp  sp  sp
0000120   sp  sp  sp  sp  sp  sp  sp  sp  sp  sp  sp  sp  sp  sp  sp  sp
0000140   sp  sp  nl   T   h   i   s  sp  sp  sp  sp   i   s  sp  sp  sp
0000160   sp   a  sp  sp  sp  sp   t   e   x   t  sp  sp  sp  sp   w   i
0000200    t   h  sp  sp  sp  sp   a  ht   t   a   b   .  nl            
0000215
```

Reply:
@ax3l Are you shure that on the machine running travis CI, the quatation markes from [line 65](https://github.com/ComputationalRadiationPhysics/picongpu/blob/3b10f00ee45fa9e9203214867aafa3ab3f19f54c/test/hasEOLwhiteSpace#L65) are printed? 

If not, another solution than the one proposed by me would be to use backslashes before the  quotation marks.

Reply:
> but when executing echo "eval \${REPLACE} $i" from here

do not execute from *there*. All the quoting is only added to print copy-pasteable output to the console and travis-ci output. That's what we then copy from.

Reply:
Now you are pointing to the hasEOL script while talking about the hasTabs script. they are different.

Reply:
Just tested again with macOS HighSierra and Bash 3.2.57 and this works.